python/ray/data/_internal/execution/streaming_executor.py

```diff
@@ -79,16 +99,110 @@ def _scheduling_loop_step(self, topology: Topology) -> bool:
             True if we should continue running the scheduling loop.
         """
 
+        if DEBUG_TRACE_SCHEDULING:
+            logger.info("Scheduling loop step...")
+
         # Note: calling process_completed_tasks() is expensive since it incurs
         # ray.wait() overhead, so make sure to allow multiple dispatch per call for
         # greater parallelism.
         process_completed_tasks(topology)
+        for op_state in topology.values():
+            op_state.refresh_progress_bar()
 
         # Dispatch as many operators as we can for completed tasks.
-        op = select_operator_to_run(topology)
+        limits = self._get_or_refresh_resource_limits()
+        cur_usage = self._get_current_usage(topology)
+        self._report_current_usage(cur_usage, limits)
+        op = select_operator_to_run(topology, cur_usage, limits)
         while op is not None:
+            if DEBUG_TRACE_SCHEDULING:
+                _debug_dump_topology(topology)
             topology[op].dispatch_next_task()
-            op = select_operator_to_run(topology)
+            cur_usage = self._get_current_usage(topology)
+            op = select_operator_to_run(topology, cur_usage, limits)
 
         # Keep going until all operators run to completion.
         return not all(op.completed() for op in topology)
+
+    def _get_or_refresh_resource_limits(self) -> ExecutionResources:
+        """Return concrete limits for use at the current time.
+
+        This method autodetects any unspecified execution resource limits based on the
+        current cluster size, refreshing these values periodically to support cluster
+        autoscaling.
+        """
+        base = self._options.resource_limits
+        cluster = ray.cluster_resources()
+        return ExecutionResources(
+            cpu=base.cpu if base.cpu is not None else cluster.get("CPU", 0.0),
```

Comment:
Do we need to deduct the resources reserved by placement groups?

Reply:
We probably have to in the future. This isn't trivial though, so for now I'm not doing any smart auto-config.